readme.md
# data

 - [bash](#bash) (288)
- [powershell](#powershell) (25)
- [python](#python) (10)
- [sql](#sql) (3)
- [cmd](#cmd) (2)
- [yaml](#yaml) (1)
- [vim](#vim) (1)
- [plain](#plain) (1)
- [javascript](#javascript) (1)
- [config](#config) (1)
## bash
- [340](data/340): && and || - compare strings using logical operators
- [339](data/339): IFS - word splitting observed in for loops
- [338](data/338): cp - brace expansion to copy/backup file
- [337](data/337): () - subshell and variable assignment
- [336](data/336): Useless use of cat
- [335](data/335): jq - convert json to environment variables
- [334](data/334): jq - extract keys/values from object
- [333](data/333): lvm - increase capacity of partition
- [332](data/332): ${0%/*} - get directory of running script
- [330](data/330): curl - display information after transfer
- [324](data/324): awk - remove duplicates without sorting
- [323](data/323): ignore carriage return in windows 
- [322](data/322): python - command line pretty print json
- [321](data/321): gitlab - create project 
- [320](data/320): jq - convert json to csv
- [319](data/319): grep - display lines in file1 that aren't in file2 (similar to comm)
- [318](data/318): declare - list all function names
- [317](data/317): nmap - view open ports
- [316](data/316): csplit - split text in file on empty lines
- [315](data/315): rpm - get all installed doc files for packages
- [314](data/314): xargs - pretty print process environment
- [313](data/313): nmap - scan port
- [312](data/312): jq - get unique items based on key
- [311](data/311): git - get remote branches
- [310](data/310): git log - get commit hash list
- [309](data/309): git-gc - cleanup unnecessary files and optimize the local repository
- [308](data/308): github api - render a markdown document as an HTML page 
- [307](data/307): fuser - close tcp ports 
- [306](data/306): openssl - create and verify CSR 
- [304](data/304): unset variables beginning with 
- [301](data/301): find - search files modified in last 24 hours
- [300](data/300): powershell - get all Get- commands
- [297](data/297): curl - search dictionaries 
- [296](data/296): wget2 - get site stats and download files
- [295](data/295): jq - target objects in array using key
- [292](data/292): naive string encryption - scramble numbers in string to hide sensitive data
- [291](data/291): netsh - find the processId for the registered urls 
- [290](data/290): netsh - Lists all defined aliases
- [289](data/289): pidstat - display global page faults and memory statistics
- [288](data/288): get docker gateway
- [286](data/286): nmap - list open TCP ports 
- [284](data/284): systemctl - view status for every loaded service on the system
- [283](data/283): journalctl - get entries for service
- [282](data/282): ps - find service using the most memory
- [281](data/281): systemd - files and tools that manage the Linux startup sequence
- [280](data/280): openssl - generate a set of prime numbers
- [279](data/279): openssl - MD5, SHA1, and SHA256 digest
- [278](data/278): openssl - extract information from a certificate
- [277](data/277): openssl - list high encryption ciphers using the AES algorithm
- [275](data/275): sssctl - get command usage from help
- [274](data/274): jq - get 
- [271](data/271): sudo - show root entries only 
- [270](data/270): awk - split string into maxtrix-like table
- [269](data/269): sed - prepending and appending to file/line
- [268](data/268): vault - get openapi specs
- [267](data/267): openssl - for versions >1.1.1
- [265](data/265): prometheus - curl metrics
- [264](data/264): ls - sortable ls
- [263](data/263): vault - enable and write userpass
- [262](data/262): get environment names matching prefix
- [260](data/260): Get the parent directory of where this script is.
- [259](data/259): investigate sockets & processes
- [258](data/258): vault - list secrets in path
- [256](data/256): curl - renew vault token
- [255](data/255): vault - update secret with file
- [254](data/254): find directories that do not have sticky bit for group
- [253](data/253): git - get top level directory
- [252](data/252): prettier - alias for stdin formatting for xml files
- [251](data/251): powershell - run noprofile command in shell (bash)
- [250](data/250): git - remove commits matching pattern in message
- [248](data/248): ansible-vault - encrypt string with vault id
- [247](data/247): uuidgen - create a new UUID value
- [246](data/246): find files created in the last hour
- [245](data/245): systemd-escape - escape strings 
- [243](data/243): openssl salted encryption/decryption
- [242](data/242): openssl crypt-style password hash
- [241](data/241): openssl benchmark
- [240](data/240): stat - display file properties
- [239](data/239): systemctl examples
- [238](data/238): openssl - verify checksum
- [237](data/237): print statistics of a process
- [236](data/236): mpstat - processors related statistics.
- [235](data/235): read network interface statistics
- [234](data/234): procinfo: display all system status gathered from /proc
- [233](data/233): show most expensive CPU process
- [232](data/232): gnu tools
- [231](data/231): list block devices
- [230](data/230): examples illustrating substring expansion
- [229](data/229): systemctl - Get enabled services
- [228](data/228): get non default values for ansible config
- [227](data/227): disable ldap login for user
- [226](data/226): restart failed services
- [225](data/225): kill all processes for user
- [224](data/224): zip files in current directory
- [223](data/223): Display the running processes of a container
- [222](data/222): nstat (better way of getting network statistics)
- [221](data/221): systemd-cgls: recursively show control group contents.
- [220](data/220): Gather listening ports
- [219](data/219): map view of pids owned by services
- [218](data/218): clean docker (to factory)
- [217](data/217): Delete Orphaned Docker Volumes
- [216](data/216): docker root systemd drop-in service
- [215](data/215): cygwin - open terminal here
- [214](data/214): find files created on this date
- [213](data/213): cleanup leftover open files
- [212](data/212): awx - get token
- [211](data/211): du - aggregate directory size
- [209](data/209): check memory usage per process
- [208](data/208): date unix timestamp to format
- [207](data/207): build json using jq
- [206](data/206): create csv from json (jq)
- [205](data/205): ps do not print columns
- [204](data/204): only get pid
- [203](data/203): os detection
- [202](data/202): get name server
- [201](data/201): Check client certificate
- [200](data/200): Cleaning up the VS Code Server on the remote
- [199](data/199): get process pid and parent id
- [198](data/198): start ssh agent on login
- [196](data/196): env reduced environment
- [194](data/194): util linux
- [191](data/191): clear screen
- [189](data/189): list files with time style & hides groups
- [186](data/186): zip directory
- [185](data/185): random variable
- [184](data/184): date - get N days ago
- [182](data/182): type - display information about command
- [181](data/181): git diff side by side
- [180](data/180): List Network Facing Services
- [179](data/179): sendmail template
- [178](data/178): sshpass - read ssh password from file
- [177](data/177): docker - resolve no space on device
- [176](data/176): list files and sort by size
- [175](data/175): Getting Server Certificate
- [174](data/174): curl vault secrets
- [172](data/172): get git symlink files
- [169](data/169): set existing variables only
- [168](data/168): delete null value
- [167](data/167): ansible-doc snippet
- [166](data/166): Get commits that have deleted files
- [164](data/164): proc environ - format for readability
- [163](data/163): comm - find common or distinct lines between files
- [162](data/162): sort top command by memory
- [161](data/161): rbind mount docker
- [160](data/160): find js files in src/
- [159](data/159): Check and assign variables
- [158](data/158): yum - get installed security updates
- [157](data/157): ssh-keygen - generate public key from prviate key
- [156](data/156): docker history
- [155](data/155): yum list docker-engine --showduplicates
- [154](data/154): umask - get umask for current process
- [153](data/153): getfactl - get user, group, and perm 
- [152](data/152): add user to group
- [151](data/151): realpath - get relative paths
- [148](data/148): Print border
- [147](data/147): get latest kernel version
- [146](data/146): awx curl
- [145](data/145): system boot menu entries
- [144](data/144): get latest kernel
- [143](data/143): get installed kernels
- [142](data/142): awx job_template
- [139](data/139): Uptime of a process
- [138](data/138): sha256
- [137](data/137): get ip address
- [132](data/132): generate random password openssl
- [131](data/131): python to encode and decode
- [130](data/130): find world writable files
- [129](data/129): password generator
- [128](data/128): pmem (vscode mimics ps)
- [127](data/127): remove trailing whitespace
- [126](data/126): /proc files
- [125](data/125): print files with timestamp
- [124](data/124): awk - remove duplicates without sorting
- [123](data/123): Use awk to display text and insert blank line before
- [122](data/122): repeat string N times
- [121](data/121): get all processes for user
- [120](data/120): get arguments after first
- [119](data/119): get runas user
- [118](data/118): string substring
- [117](data/117): string slicing
- [116](data/116): set color
- [115](data/115): ss - investigate sockets
- [114](data/114): sqlplus - set password
- [113](data/113): awk - get value only
- [112](data/112): view process id
- [111](data/111): exec cmd
- [110](data/110): generate random password
- [108](data/108): run sudosu
- [107](data/107): get allowed sudo commands
- [106](data/106): get git status for all worktree repos in git
- [105](data/105): find zombie processes
- [104](data/104): extract in a different directory
- [103](data/103): compress directory using tar
- [102](data/102): get top memory/cpu hog
- [101](data/101): Get systemctl property
- [100](data/100): docker volume - inspect all
- [99](data/99): generating crypto quote
- [98](data/98): git diff tags
- [97](data/97): sed - only print after matching pattern
- [96](data/96): sed - delete rows after pattern 
- [95](data/95): sed replace files
- [94](data/94): sort .bash_history
- [93](data/93): self sign cert (generate private key, csr, and crt)
- [92](data/92): git pager
- [89](data/89): prune unwritable files from find
- [86](data/86): psql table size
- [84](data/84): print all commands in $PATH (windows bash)
- [83](data/83): type
- [82](data/82): ansible list all hosts
- [81](data/81): encrypt/decrypt gpg
- [80](data/80): ansible - transfer a file directly to many servers:
- [79](data/79): See the list of available plugins
- [78](data/78): generate ed25519 private key
- [77](data/77): rsync
- [76](data/76): create csr request
- [75](data/75): print markdown docs not in node_modules
- [74](data/74): get most used commands
- [73](data/73): print commands from sudo log
- [72](data/72): view UDP and TCP ports
- [71](data/71): output all yml
- [70](data/70): encrypt string with ansible-vault
- [69](data/69): list all ip rules
- [68](data/68): get matching prefix
- [67](data/67): greedy *
- [66](data/66): hack for exporting env variables
- [65](data/65): get modification date for files in directory
- [64](data/64): view workstation details
- [63](data/63): Recursively get full path name
- [62](data/62): coreutils
- [61](data/61): get sorted size of directories in current working directory
- [60](data/60): Resolve warning that bridge-nf-call-iptables is disabled
- [59](data/59): Get size of mounted filesystems
- [58](data/58): alias for easy column extraction
- [57](data/57): view service account owner
- [56](data/56): fetch all git tags
- [55](data/55): get free space on directory
- [54](data/54): get free space in directory
- [53](data/53): hack for unsourcing env variables
- [52](data/52): echo shell commands as they are executed
- [51](data/51): print all the columns after a particular number using awk
- [50](data/50): set timeout for script 
- [49](data/49): git make files executable
- [48](data/48): curl yum repo
- [47](data/47): set env variables
- [46](data/46): umask create file with permissions
- [45](data/45): pip show for all pip installed
- [44](data/44): capturing group grep
- [43](data/43): docker history dockerfile
- [42](data/42): docker save images
- [41](data/41): pip install requirements one line
- [40](data/40): print env variables that start with S
- [39](data/39): Parse a .env (dotenv) file directly using bash
- [38](data/38): extract title from html
- [37](data/37): rsync with exclude
- [36](data/36): view signed certs
- [35](data/35): hash directory 
- [34](data/34): print name + content
- [32](data/32): diff project structures for git branches
- [31](data/31): download & save as json
- [30](data/30): a fancier lsof -i
- [29](data/29): get list of most used commands
- [28](data/28): Graph # of connections for each hosts.
- [27](data/27): ps examples
- [26](data/26): print filename + content of sudo privileges
- [25](data/25): create html toc
- [24](data/24): urldecode
- [23](data/23): recursively download files from web
- [22](data/22): creating a user with pass
- [21](data/21): sending email
- [20](data/20): Get active internet connections and domain sockets
- [19](data/19): unwrap text file (useful for logs)
- [18](data/18): clear pycache
- [17](data/17): print = across the screen
- [16](data/16): Update multiple repos (no rev)
- [15](data/15): view file structure (tree)
- [14](data/14): List full paths for recursive files with grep and awk
- [13](data/13): Updating Multiple Repos With One Command
- [11](data/11): append group to a user
- [10](data/10): view os-release
- [9](data/9): cf
- [8](data/8): rpm
- [7](data/7): test openssl connection
- [6](data/6): display information about the CPU architecture
- [5](data/5): start firewalld
- [4](data/4): find and kill processes
- [3](data/3): host
- [2](data/2): pstree
- [1](data/1): generate RSA key

[[Return to top]](#data)

## powershell
- [303](data/303): Get-WindowsUpdateLog - get update
- [302](data/302): Finds local admin password and password expiration timestamp for given computer
- [299](data/299): Get-NetAdapter
- [298](data/298): Get-WmiObject - get network adapter config
- [294](data/294): WindowsIdentity- get current user in powershell
- [293](data/293): env - get all environment items in powershell
- [287](data/287): Get-ADGroupMember
- [276](data/276): Powershell - set variables to get dynamic ADUser
- [273](data/273): read powershell history and trim leading whitespace
- [272](data/272): get-help - show all loaded cmdlets, functions, and modules in powershell
- [266](data/266): get-adgroup - get all members for AD group
- [249](data/249): password last reset
- [195](data/195): List available powershell commands
- [188](data/188): Get-WmiObject - get computer system properties
- [171](data/171): Get AD User from another domain (ldap)
- [170](data/170): get AD domain controller (amer.example.com)
- [141](data/141): get laptop service tag
- [140](data/140): Get all powershell modules
- [136](data/136): Get group members
- [134](data/134): Powershell get calculated property
- [133](data/133): Get service accounts owned by user and display in table
- [91](data/91): Get services with search string
- [90](data/90): get all available data about a process
- [88](data/88): get ldap group related to ADUser
- [87](data/87): get AD User (powershell)

[[Return to top]](#data)

## python
- [331](data/331): python - select keys in json to print
- [329](data/329): python module calendar
- [328](data/328): sysconfig - provide access to Python's configuration information
- [327](data/327): cProfile - Performance Profiling Python One-Liner
- [326](data/326): python - csv to json
- [325](data/325): python oneliner - print columns like awk
- [261](data/261): convert json to yaml from stdin
- [257](data/257): venv 
- [187](data/187): Start a static HTTP server
- [135](data/135): awx-manage shell_plus

[[Return to top]](#data)

## sql
- [193](data/193): psql: cast json and pretty print
- [192](data/192): postgresql list and order tables by size
- [190](data/190): get list of tables and number of rows (postgres)

[[Return to top]](#data)

## cmd
- [305](data/305): fsutil resource info
- [183](data/183): bat file for opening cygwin in current directory

[[Return to top]](#data)

## yaml
- [244](data/244): ansible playbook variables

[[Return to top]](#data)

## vim
- [285](data/285): vim - display all possible runtime config settings for vim

[[Return to top]](#data)

## plain
- [149](data/149): ansible-playbook in crontab

[[Return to top]](#data)

## javascript
- [173](data/173): devtools: get all css items

[[Return to top]](#data)

## config
- [197](data/197): enable ssh controlmaster - ssh sessions

[[Return to top]](#data)



changelog.md
# Changelog

## 2022-05-26

- Modified #327: cProfile - Performance Profiling Python One-Liner
- Modified #325: python oneliner - print columns like awk
- Modified #324: awk - remove duplicates without sorting
- Added #328: sysconfig - provide access to Python's configuration information
- Added #327: cProfile - Performance Profiling Python One-Liner
- Added #326: python - csv to json

22:04PM
- Modified #327: cProfile - Performance Profiling Python One-Liner
- Modified #329: python module calendar
- Modified #330: curl - display information after transfer
- Modified #331: python - select k
- Added #330: curl - display information after transfer
- Added #331: python - select k


public/script.js
const $$ = query => Array.from(document.querySelectorAll(query))

function getSnippetsUrl() {
  return 'https://raw.githubusercontent.com/nntrn/sshell/data/snippets.json'
}

function removeHash() {
  history.replaceState({}, document.title, window.location.href.split('#')[0])
}

function groupByKey(obj, key) {
  return obj.reduce((group, product) => {
    const value = product[key]
    group[value] = group[value] ?? []
    group[value].push(product)
    return group
  }, {})
}

function escapeHtml(s) {
  var ENTITY_MAP = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#39;',
    '/': '&#x2F;',
  }
  return ('' + s).replace(/[&<>"'/]/g, function (s) {
    return ENTITY_MAP[s]
  })
}

function buildSnippets(res) {
  Object.entries(groupByKey(res, 'language')).forEach(langGrp => {
    const group = Object.assign(document.createElement('div'), {
      className: 'group',
      id: langGrp[0],
    })

    group.appendChild(
      Object.assign(document.createElement('h3'), {
        textContent: langGrp[0],
      })
    )

    document.querySelector('#toc').appendChild(
      Object.assign(document.createElement('a'), {
        className: 'link',
        href: `#${langGrp[0]}`,
        textContent: langGrp[0],
      })
    )

    langGrp[1].forEach(({ output, ...e }) => {
      const itemjson = JSON.stringify(e, null, 2)
      const codeText = escapeHtml([e.cli].flat(2).join('\n'))
      group.appendChild(
        Object.assign(document.createElement('div'), {
          className: 'item',
          id: `data-${e.id}`,
          innerHTML: [
            `<h4><a href="#${e.id}" onclick="showPopup(${e.id})">${e.title}</a></h4>`,
            `<div class="code" style="user-select:all">${codeText}</div>`,
            `<div class="filter" style="display:none">${itemjson}</div>`,
          ].join('\n'),
        })
      )
    })
    document.querySelector('#sshell').appendChild(group)
  })
}

function searchItems(string) {
  Array.from(document.querySelectorAll('.item .filter')).forEach(e => {
    if (e.textContent.includes(string)) {
      e.parentElement.classList.remove('hidden')
    } else {
      e.parentElement.classList.add('hidden')
    }
  })
}

function unhideAllItems(e) {
  console.log(e)
  Array.from(document.querySelectorAll('.item .hidden')).forEach(e => {
    e.classList.remove('hidden')
  })
}

function eventFilter(e) {
  const keycodes = {
    enter: 13,
    space: 32,
    backspace: 8,
  }
  if (Object.values(keycodes).includes(e.keyCode)) {
    console.log(e.keyCode, e.key)
    const searchString = e.target.value.toLowerCase().trim()
    searchItems(searchString)
  }
}

function hidePopup() {
  document.querySelector('#popup').classList.remove('show')
  Array.from(document.querySelectorAll('#popup [data-key]')).forEach(e => (e.textContent = ''))
  removeHash()
}

function getLocationId() {
  return location.hash.slice(1)
}

function editPopupText(obj) {
  $$('#popup [data-key]').forEach(e => {
    if (e.nodeName.toLowerCase() === 'a') {
      e.href = obj[e.dataset.key]
    }
    e.textContent = [obj[e.dataset.key]].flat(2).join('\n')
  })
}

function showPopup(id = getLocationId()) {
  editPopupText(JSON.parse(document.querySelector(`#data-${id} .filter`).textContent))
  document.querySelector('#popup').classList.add('show')
}

document.querySelector('#searchbar').addEventListener('keyup', eventFilter)

document.querySelector('#searchbar').addEventListener('click', unhideAllItems)

document.onkeydown = function (evt) {
  if (evt.keyCode === 27) {
    hidePopup()
  }
}

window.addEventListener('load', function () {
  fetch(getSnippetsUrl())
    .then(e => e.json())
    .then(res => buildSnippets(res))
    .then(e => {
      if (location.hash) {
        showPopup()
      }
    })
    .catch(err => console.log(err))
})
